MapboxNavigation/NavigationMapView.swift

```diff
@@ -427,36 +426,96 @@ open class NavigationMapView: MGLMapView {
+        if let activeRoute = waypointTest?.multipointRoutes.first ,let waypointToRemove = waypointTest?.waypoints.first { //test passes
+            return requestNew(route: activeRoute, without: waypointToRemove)
+        } else if let routes = self.routes(closeTo: sender) {
+            guard let selectedRoute = routes.first else { return }
+            select(route: selectedRoute)
+        }
         
-        let closestRoute = routes.filter {
-            guard let coords = $0.coordinates else { return false }
-            return coords.count > 1
-        }.min { (left, right) -> Bool in
-            let leftDistance = Polyline(left.coordinates!).closestCoordinate(to: tapCoordinate)!.distance
-            let rightDistance = Polyline(right.coordinates!).closestCoordinate(to: tapCoordinate)!.distance
+    }
+    
+    //TODO: Change to point-based distance calculation
+    private func waypoints(on routes: [Route], closeTo tap: UITapGestureRecognizer) -> (multipointRoutes: [Route], waypoints: [Waypoint])? {
+        guard let tapCoordinate = tap.coordinate(in: self), let tapPoint = tap.point else { return nil }
+        let waypointThreshold: CGFloat = 100 //FIXME: Class Constant
+        let multipointRoutes = routes.filter { $0.routeOptions.waypoints.count >= 3}
+        guard multipointRoutes.count > 0 else { return nil }
+        let waypoints = multipointRoutes.flatMap({$0.routeOptions.waypoints})
+        
+        //lets sort the array in order of closest to tap
+        let closest = waypoints.sorted { (left, right) -> Bool in
+            let leftDistance = left.coordinate.distance(to: tapCoordinate)
+            let rightDistance = right.coordinate.distance(to: tapCoordinate)
             return leftDistance < rightDistance
         }
         
-        if let closestRoute = closestRoute, let closestCoordinate = Polyline(closestRoute.coordinates!).closestCoordinate(to: tapCoordinate), let routeIndex = routes.index(where: { $0 ==  closestRoute}) {
-            let closestPoint = self.convert(closestCoordinate.coordinate, toPointTo: self)
-            let tapDistanceFromClosestRoute = closestPoint.distance(to: tapPoint)
-            if tapDistanceFromClosestRoute <= maximumTapDistanceToSelectAltRoute {
-                let selectedRoute = routes[routeIndex]
-                self.routes?.remove(at: routeIndex)
-                self.routes?.insert(selectedRoute, at: 0)
-                self.showRoutes(self.routes!)
-                navigationMapDelegate?.navigationMapView?(self, didTap: selectedRoute)
-            }
+        //lets filter to see which ones are under threshold
+        let candidates = closest.filter({
+            let coordinatePoint = self.convert($0.coordinate, toPointTo: self)
+            return coordinatePoint.distance(to: tapPoint) < waypointThreshold
+        })
+        
+        return (multipointRoutes, candidates)
+    }
+    
+    private func routes(closeTo tap: UITapGestureRecognizer) -> [Route]? {
```

Comment:
This method should take a point (in screen coordinates), rather than a gesture recognizer. That’ll make it useful for other purposes too.

Reply:
The problem with this is that we do our distance calculation based upon `CGPoint` distance, not `CLLocationDistance`. I need the recognizer so I can convert coordinates back into screen geometry, and then test if the tap was within 50 points of the closest route coordinate.

Reply:
Agreed, my suggestion is to have the method taking a CGPoint, not a CLLocationCoordinate2D.

Reply:
Apologies, I misread. Will Do.